reddit/autorefresh.go

```diff
@@ -0,0 +1,48 @@
+package reddit
+
+import (
+	"log"
+	"time"
+
+	"golang.org/x/net/context"
+	"golang.org/x/oauth2"
+)
+
+type autoRefresh struct {
+	refreshTokenTimer time.Duration
+	cancelAutorefresh context.CancelFunc
+	context           context.Context
+}
+
+func newAutoRefresh() *autoRefresh {
+	ctx := context.Background()
+	ctx, cancel := context.WithCancel(ctx)
+
+	return &autoRefresh{
+		cancelAutorefresh: cancel,
+		context:           ctx,
+	}
+}
+
+func (ar *autoRefresh) setRefreshTimerFromToken(token *oauth2.Token) {
+	ar.refreshTokenTimer = token.Expiry.Sub(time.Now())
+}
+
+func (ar *autoRefresh) autoRefresh(a *appClient) {
+	timer := time.NewTimer(ar.refreshTokenTimer)
```

Comment:
Instead of a timer, why not do this synchronously as a strategy for remediation of a failed request?

It is simpler to do it that way and the cost of an extra round trip to refresh the token is not bad to pay; latency against Reddit naturally jitters more than that.

Reply:
It could also make sense to preemptively check for token expiry and refresh ahead of time if it is expired.

Reply:
I think it is actually simpler doing it that way. This prevent doing any checks whatsoever or re-trigger any failed request.
The timer use the refresh time contained in the response to access_token endpoint. I can make the timer refresh some seconds before the token expire to avoid any failed request or latency from Reddit. Not optimal but costs nothing.

Reply:
Doing this concurrently introduces the complexities of races, cancellation, and synchronization.

I don’t want that if it isn’t necessary. If I wrote this library again, there would be no channels or go routines in it.

Reply:
I agree with you, in general it adds complexity. But this goroutine is quite simple, doing its job on its own without adding any complexity, no races, no synchronization.

To me this kind of feature is perfect with the use of a goroutine.

Reply:
Consider:

* A user creates a bot, this goroutine is running, and then they close the bot. Who cleans up the goroutine?

* The goroutine doesn't get scheduled precisely, and the user sees an error because this ran just a little too late.

* The goroutine fails. How does it communicate back to the main thread that another one needs to be started?

There are more; these are just off the top of my head. You could write code to solve all of these cases, but we can't be sure we thought of them all without a great deal of effort.

The end result would be less code than this PR, more stable, and more resilient to code changes if this logic were just handled synchronously in the normal request flow.

Reply:
Ah yes! This being a package and not a binary, I should have implemented goroutine cancel mechanic when a bot is closed. My bad.

The goroutine has no way to fail in this state. If it does, this means panic.

If I understood correctly, what you propose is to synchronously refresh the token only when it needs to. After a failed request or before a request.

`After a failed request` means we have to re-trigger any failed requests. Too complicated.

`Before a request`: Check if expired or soon to be before a request would be easy. Pretty much recall authorize() with condition, adding some lag to the request. 
But what if the request to access_token fails ? We do it again until it works ? What would be a good non-incapacitating behavior ?

For example, Reddit server is busy. Beforehand we reduce the wait timer to a certain amount of time (45min instead of an 1hour). If the call to access_token fails, we can make the goroutines call itself every "x" seconds until the token is refreshed. This allows more flexibility and, if needed, to add more logic inside the goroutine to make sure token will be always fresh without touching or slowing the requests flow.

I never had any problem using simple goroutines like this one but if by experience you had some cases that made you want to never use goroutines again, please share if you can/want :-)

Reply:
The problem you describe about what to do when getting a refresh token fails applies invariably to the synchronous solution and this one as well.

In addition to the complexities to this approach I listed above which I think remain unsolved:

* It is not clear to me that this is free of data-races. There are two goroutines using `appClient` and nothing in place to synchronize them.

* This is looping by recursion, and Go does not have tail calls, so you are leaking memory in this goroutine.

Doing it synchronously adds a branch here: https://github.com/turnage/graw/blob/master/reddit/appclient.go#L25 . Re-firing the request is easy because you still hold the request. Sending a request doesn't mutate it.

I think I have sufficiently explained my position on this. At this point I'll just say I'm not going to merge the code with an extra goroutine.